webapp/src/ts/services/transitions/muting.transition.ts

```diff
@@ -0,0 +1,440 @@
+import { Injectable } from '@angular/core';
+
+import { DbService } from '@mm-services/db.service';
+import { LineageModelGeneratorService } from '@mm-services/lineage-model-generator.service';
+import { ContactMutedService } from '@mm-services/contact-muted.service';
+import { ContactTypesService } from '@mm-services/contact-types.service';
+import { TransitionInterface } from '@mm-services/transitions/transition';
+import { ValidationService } from '@mm-services/validation.service';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class MutingTransition implements TransitionInterface {
+  constructor(
+    private dbService:DbService,
+    private lineageModelGeneratorService:LineageModelGeneratorService,
+    private contactMutedService:ContactMutedService,
+    private contactTypesService:ContactTypesService,
+    private validationService:ValidationService,
+  ) { }
+
+  readonly name = 'muting';
+
+  private transitionConfig;
+  private inited;
+  private readonly CONFIG_NAME = 'muting';
+  private readonly MUTE_PROPERTY = 'mute_forms';
+  private readonly UNMUTE_PROPERTY = 'unmute_forms';
+  private readonly OFFLINE = 'offline';
+
+  private getConfig(settings = {}) {
+    return settings[this.CONFIG_NAME] || {};
+  }
+
+  private getMutingForms() {
+    return this.transitionConfig?.[this.MUTE_PROPERTY] || [];
+  }
+
+  getUnmutingForms(settings?) {
+    const config = settings ? this.getConfig(settings) : this.transitionConfig;
+    return config?.[this.UNMUTE_PROPERTY] || [];
+  }
+
+  init(settings) {
+    this.transitionConfig = this.getConfig(settings);
+
+    const mutingForms = this.getMutingForms();
+    if (!mutingForms || !Array.isArray(mutingForms) || !mutingForms.length) {
+      console.warn(
+        `Configuration error. Config must define have a '${this.CONFIG_NAME}.${this.MUTE_PROPERTY}' array defined.`
+      );
+      this.inited = false;
+    } else {
+      this.inited = true;
+    }
+
+    return this.inited;
+  }
+
+  private isMuteForm(form) {
+    return !!form && this.getMutingForms().includes(form);
+  }
+
+  isUnmuteForm(form, settings?) {
+    return !!form && this.getUnmutingForms(settings).includes(form);
+  }
+
+  /**
+   * Returns whether a document is a muting or unmuting report that should be processed.
+   * We only process new reports. The muting transition should not run when reports are edited.
+   * @param {Doc} doc - the doc to check
+   * @returns {Boolean} - whether this is a new muting or unmuting report
+   * @private
+   */
+  private isRelevantReport(doc) {
+    // exclude docs that are not reports and existent reports.
```

Comment:
Why are we ignoring existent reports? What happens if someone edits a form to mute a different person?

Reply:
Nothing. Just like sentinel, transitions only run once.